srml/babe/src/lib.rs

```diff
@@ -151,16 +170,38 @@ decl_module! {
 				.iter()
 				.filter_map(|s| s.as_pre_runtime())
 				.filter_map(|(id, mut data)| if id == BABE_ENGINE_ID {
-					<[u8; VRF_OUTPUT_LENGTH]>::decode(&mut data)
+					<(u64, [u8; VRF_OUTPUT_LENGTH])>::decode(&mut data)
 				} else {
 					None
 				}) {
-				Self::deposit_vrf_output(&i);
+				let slot_index = i.0
+					.checked_add(1)
+					.expect("slot numbers will never reach 2^64 in our lifetimes; qed");
+				let epoch_change_slot = slot_index
+					.checked_add(Self::slots_per_epoch())
+					.expect("slot numbers will never reach 2^64 in our lifetimes; qed");
+				if epoch_change_slot == slot_index {
+					// New epoch
+					LastSlotInEpoch::put(true)
+				}
+				return Self::deposit_vrf_output(&i.1)
 			}
 		}
 	}
 }
 
+pub fn epoch<T: Trait>() -> Epoch {
+	let authorities = Authorities::get();
+	if false && authorities.is_empty() {
+		panic!("No authorities ― network is deadlocked")
+	}
+	Epoch {
+		randomness: EpochRandomness::get(),
+		authorities: authorities,
+		slot_number: SlotsPerEpoch::get().wrapping_add(EpochIndex::get()),
```

Comment:
Is this correct? For one epoch this will always return the same slot number?

Reply:
Fixed by deleting the `slot_number` field.